File: PlayGround/LinkedList/SingleLinkList/LinkedList.py
arr = [1,4,5,6,7]

print(arr)

#list -> linkedlist
class LinkedList:
    none = None

    def __init__(self, val=0, next = none) -> None:
        assert next is LinkedList.none or isinstance(next, LinkedList)
        self.val = val
        self.next = next

def changeListToLinkedList(arr):
    if len(arr) == 1:
        return LinkedList(arr[0])
    else:
        return LinkedList(arr[0], changeListToLinkedList(arr[1:]))

def printLinkedList(linkedlist):
    head = linkedlist

    while head:
        print("->", head.val, end="")
        head = head.next

    print("\n")

linkL = changeListToLinkedList(arr)
printLinkedList(linkL)

#linkedlist reverse
def reverseLinkedList(linkedlist):
    head = linkedlist
    curr = head
    prev = None


    while curr:
        next = curr.next
        curr.next =prev
        prev = curr
        curr = next

    head = prev
    return head

linkL = reverseLinkedList(linkL)
printLinkedList(linkL)

#linkedlist -> list
def changeLinkedListToArr(head):
    arr = []
    while head:
        arr.append(head.val)
        head = head.next

    return arr

print(changeLinkedListToArr(linkL))

File: PlayGround/LinkedList/SingleLinkList/LinkedList.go
package main

import "fmt"

type Node struct {
	next *Node
	key  interface{}
}

type SingleLinkedList struct {
	head *Node
}

func (S *SingleLinkedList) Insert(key interface{}) {
	singleLinkedList := &Node{
		next: S.head,
		key:  key,
	}

	S.head = singleLinkedList

	l := S.head
	for l.next != nil {
		l = l.next
	}
}

func (S *SingleLinkedList) changeListIntoLinkedList(arr []int) {
	for i := len(arr) - 1; i >= 0; i-- {
		S.Insert(arr[i])
	}
}

func (S *SingleLinkedList) PrintLinkedList() {
	list := S.head
	for list != nil {
		fmt.Printf("%+v ->", list.key)
		list = list.next
	}

	fmt.Println()
}

func PrintLinkedList(list *Node) {
	for list != nil {
		fmt.Printf("%v ->", list.key)
		list = list.next
	}

	fmt.Println()
}

func (S *SingleLinkedList) Reverse() {
	curr := S.head
	var prev *Node

	for curr != nil {
		next := curr.next
		curr.next = prev
		prev = curr
		curr = next
	}

	S.head = prev

	PrintLinkedList(S.head)
}

func (S *SingleLinkedList) changeLinkedListToArray() []int {
	arr := []int{}
	list := S.head

	for list != nil {
		arr = append(arr, list.key.(int))
		list = list.next
	}

	return arr
}

func main() {
	// arr -> linked List
	arr := []int{1, 4, 5, 6, 7}
	link := SingleLinkedList{}

	fmt.Println(arr)
	link.changeListIntoLinkedList(arr)
	link.PrintLinkedList()

	// reverse
	link.Reverse()

	// linked list -> arr
	result := link.changeLinkedListToArray()
	fmt.Println(result)

}
